airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/BigQueryDestination.kt

```diff
@@ -83,23 +96,189 @@ class BigQueryDestination : BaseConnector(), Destination {
                     .setUseLegacySql(false)
                     .build()
 
+            val result = BigQueryUtils.executeQuery(bigquery, queryConfig)
+
+            if (result.getLeft() == null) {
+                throw ConfigErrorException(result.right)
+            }
+
             if (UploadingMethod.GCS == uploadingMethod) {
-                val status = checkGcsPermission(config)
-                if (status!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
-                    return status
+
+                val gcsStatus = checkGcsAccessPermission(config)
+                if (gcsStatus!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
+                    return gcsStatus
+                }
+
+                // Copy a temporary dataset to confirm copy permissions are working
+                val bigQueryStatus = checkBigQueryCopyPermission(config)
+
+                if (bigQueryStatus!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
+                    return bigQueryStatus
                 }
             }
 
-            val result = BigQueryUtils.executeQuery(bigquery, queryConfig)
-            if (result.getLeft() != null) {
-                return AirbyteConnectionStatus()
-                    .withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+            return AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+        } catch (e: Exception) {
+            log.error(e) { "Check failed." }
+            throw ConfigErrorException((if (e.message != null) e.message else e.toString())!!)
+        }
+    }
+
+    /**
+     * This method performs a copy operation to copy data into a temporary table on BigQuery to
+     * check if the existing permissions are sufficient to copy data. If the permissions are not
+     * sufficient, then an exception is thrown with a message showing the missing permission
+     */
+    private fun checkBigQueryCopyPermission(config: JsonNode): AirbyteConnectionStatus? {
+
+        // TODO: Need to add a step in this method to first check permissions
```

Comment:
is this something you're planning to do as part of this PR, or a future enhancement? (if future enhancement: can you write an issue for it? https://github.com/airbytehq/airbyte-internal-issues/issues/new/)

Reply:
I have created a new issue to track the work needed for proactively checking required BigQuery permissions: https://github.com/airbytehq/airbyte-internal-issues/issues/8888

Reply:
nice. I think it's best practice to link to that issue in a code comment, just so we can remember it in the future

Reply:
sure, I have added a link for the issue in the TODO comment